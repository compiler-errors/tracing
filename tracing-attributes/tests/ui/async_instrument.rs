#![allow(unreachable_code)]

#[tracing::instrument]
async fn unit() {
    ""
}

#[tracing::instrument]
async fn simple_mismatch() -> String {
    ""
}

// FIXME: this span is still pretty poor
#[tracing::instrument]
async fn opaque_unsatisfied() -> impl std::fmt::Display {
    ("",)
}

struct Wrapper<T>(T);

#[tracing::instrument]
async fn mismatch_with_opaque() -> Wrapper<impl std::fmt::Display> {
    ""
}

fn main() {
    let _ = unit();
    let _ = simple_mismatch();
    let _ = opaque_unsatisfied();
    let _ = mismatch_with_opaque();
}
