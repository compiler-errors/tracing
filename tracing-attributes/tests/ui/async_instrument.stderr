error[E0308]: mismatched types
 --> tests/ui/async_instrument.rs:5:5
  |
5 |     ""
  |     ^^ expected `()`, found `&str`

error[E0308]: mismatched types
  --> tests/ui/async_instrument.rs:10:5
   |
10 |     ""
   |     ^^- help: try using a conversion method: `.to_string()`
   |     |
   |     expected struct `String`, found `&str`
   |
note: return type inferred to be `String` here
  --> tests/ui/async_instrument.rs:9:31
   |
9  | async fn simple_mismatch() -> String {
   |                               ^^^^^^

error[E0277]: `(&str,)` doesn't implement `std::fmt::Display`
  --> tests/ui/async_instrument.rs:14:1
   |
14 | #[tracing::instrument]
   | ^^^^^^^^^^^^^^^^^^^^^^ `(&str,)` cannot be formatted with the default formatter
   |
   = help: the trait `std::fmt::Display` is not implemented for `(&str,)`
   = note: in format strings you may be able to use `{:?}` (or {:#?} for pretty-print) instead
   = note: this error originates in the attribute macro `tracing::instrument` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: `(&str,)` doesn't implement `std::fmt::Display`
  --> tests/ui/async_instrument.rs:15:34
   |
15 | async fn opaque_unsatisfied() -> impl std::fmt::Display {
   |                                  ^^^^^^^^^^^^^^^^^^^^^^ `(&str,)` cannot be formatted with the default formatter
   |
   = help: the trait `std::fmt::Display` is not implemented for `(&str,)`
   = note: in format strings you may be able to use `{:?}` (or {:#?} for pretty-print) instead

error[E0308]: mismatched types
  --> tests/ui/async_instrument.rs:23:5
   |
23 |     ""
   |     ^^ expected struct `Wrapper`, found `&str`
   |
   = note: expected struct `Wrapper<_>`
           found reference `&'static str`
note: return type inferred to be `Wrapper<_>` here
  --> tests/ui/async_instrument.rs:22:36
   |
22 | async fn mismatch_with_opaque() -> Wrapper<impl std::fmt::Display> {
   |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: try wrapping the expression in `Wrapper`
   |
23 |     Wrapper("")
   |     ++++++++  +
